modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServices.java

```diff
@@ -581,6 +582,26 @@ public Object invoke(String mthdName, boolean srvKeepBinary, Object[] args)
 
                 Method mtd = getMethod(serviceClass, mthdName, args);
 
+                // Convert Object[] to T[] when required:
+                // Ignite loses array item types when passing arguments through GridServiceProxy.
+                for (int i = 0; i < args.length; i++) {
```

Comment:
IgniteServices has a "keep binary" flag that makes the server side NOT to deserialize parameters, thus, allowing NOT having Java classes on the server. You need to do this block only if the "keep binary" flag is not set. Otherwise it will do that Array.newInstance thing even when "keep binary = true".

Reply:
`Array.newInstance` is called only when array type does not match. `keepBinary` mode requires the same treatment as normal mode - service may take `BinaryObject[]`, so array conversion is necessary. I've added tests for both .NET->.NET and .NET->Java calls to demonstrate this.